hs/src/Reach/Proto.hs

```diff
@@ -0,0 +1,190 @@
+--
+-- TODO much of the following ought to be renamed
+-- TODO "AppT" over `say`, `listen`, and `cmd` request helpers
+-- TODO easy thread-safety
+-- TODO prolong client timeouts
+
+stubPort :: Int
+stubPort = 8123
+
+data EventStream deriving Typeable
+
+data FDescOut'
+  = Stdout'
+  | Stderr'
+  | Stdboth'
+  deriving Eq
+
+instance FromHttpApiData FDescOut' where
+  parseUrlPiece = \case
+    "1" -> Right Stdout'
+    "2" -> Right Stderr'
+    "3" -> Right Stdboth'
+    a -> Left $ "Invalid fd: " <> a
+
+instance ToHttpApiData FDescOut' where
+  toUrlPiece = \case
+    Stdout'  -> "1"
+    Stderr'  -> "2"
+    Stdboth' -> "3"
+
+data FDescOut
+  = Stdout Integer BL.ByteString
+  | Stderr Integer BL.ByteString
+  | Keepalive
+
+type Project = M.Map FilePath Text -- path:hash
+
+data Req = Req
+  { req_args  :: [Text]
+  , req_files :: Maybe Project
+  , req_env   :: M.Map Text Text
+  } deriving (Eq, Generic, FromJSON, ToJSON)
+
+data Res = Res -- TODO
+  deriving (Eq, Generic, FromJSON, ToJSON)
+
+type Pid = Text
+
+instance Accept EventStream where
+  contentType _ = "text" H.// "event-stream"
+
+class ToSSE a where
+  toSSE :: a -> BL.ByteString
+
+toSSE' :: Show a => BL.ByteString -> a -> BL.ByteString -> BL.ByteString
+toSSE' e n m
+  = "id: " <> BL.fromString (show n) <> "\n"
+ <> "event: " <> e <> "\n"
+ <> "data: " <> m <> "\n\n"
+
+instance ToSSE FDescOut where
+  toSSE = \case
+    Stdout n m -> toSSE' "stdout" n m
+    Stderr n m -> toSSE' "stderr" n m
+    Keepalive  -> ": keepalive\n\n"
+
+instance ToSSE a => MimeRender EventStream a where
+  mimeRender _ = toSSE
+
+pFDescOut :: Parsec String () FDescOut
+pFDescOut = pka <|> pout where
+  TokenParser {..} = makeTokenParser emptyDef
+  pka = (string ": keepalive\n\n" *> pure Keepalive) <* eof
+  pout = do
+    i <- string "id: " *> ((try $ char '0' *> pure 0 <* newline) <|> natural)
+    e <- string "event: "
+      *> ((try $ string "stdout" *> pure Stdout') <|> (string "stderr" *> pure Stderr'))
+      <* newline
+    m <- BL.fromString <$> (string "data: " *> manyTill anyChar (try $ string "\n\n") <* eof)
+    case e of
+      Stdout' -> pure $ Stdout i m
+      Stderr' -> pure $ Stderr i m
+      _ -> fail "Event was neither `stdout` nor `stderr`"
+
+instance MimeUnrender EventStream FDescOut where
+  mimeUnrender _ = either (Left . show) Right . runParser pFDescOut () "" . BL.toString
+
+type V0_Sync = "v0" :>
+    ( "exec" :> "reach" :> Capture "cmd" String :> ReqBody '[JSON] Req :> Post '[JSON] Res
+ :<|> "proc" :> Capture "pid" Pid :> "fd" :> "0" :> ReqBody '[PlainText] Text :> PutNoContent
+    )
+
+type V0_Stream = "v0"
+  :> "proc" :> Capture "pid" Pid
+  :> "fd"   :> Capture "fd"  FDescOut'
+  :> StreamGet NoFraming EventStream (SourceIO FDescOut)
+
+type Proto = V0_Sync :<|> V0_Stream
+
+runStubServer :: IO ()
+runStubServer = Warp.run stubPort $ serve (Proxy @Proto) ((r :<|> pin) :<|> pout) where
```

Comment:
Do we imagine that is a demonstration of how to write a server or that this is the library that will actually do it? If library, then I think this should be a bit more parameterized, like `runServer :: MonadIO m => Config -> (Req -> m Resp) -> m ()` for a suitable part of `Req`/`Resp`

Reply:
I think of this function as more of a demo, but other elements of this module are the beginnings of a library.

In real-world servers it'd be more common to break stuff in `runStubServer` out into abstracted + distinct pieces, so it naturally becomes easier to parameterize.

I think `runServer :: Config -> ...` or `c <- readConfig` will be a common pattern for all our REACHp-speakers though, yup. I agree we could probably make a helper function to capture the pattern.